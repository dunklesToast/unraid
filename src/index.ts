export * as ssh from './executors/SSH';
export * as vm from './modules/vms/vm-module';
