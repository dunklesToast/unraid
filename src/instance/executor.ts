import type { EventEmitter } from 'events';

export type IExecute = {
    command: string;
};

export type IExecuteResult = {
    stdout: string[];
    stderr: string[];
    code: number;
    signal?: unknown;
};

export type IExecuteStreamResult = {
    code: number;
    signal?: unknown;
};

export type IExecuteSimple = string;

export type CancelFunction = () => Promise<void>;

export type ExecutorConfigType = Record<string, unknown>;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export interface Executor<Config> {
    executeStream?(
        command: IExecuteSimple | IExecute
    ): Promise<[EventEmitter, CancelFunction, Promise<IExecuteStreamResult>]>;
}

export abstract class Executor<Config extends ExecutorConfigType = ExecutorConfigType> {
    readonly config: Config;

    constructor(config: Config) {
        this.config = config;
    }

    abstract execute(command: IExecuteSimple): Promise<IExecuteResult>;

    abstract execute({ command }: IExecute): Promise<IExecuteResult>;

    abstract execute(command: IExecuteSimple | IExecute): Promise<IExecuteResult>;
}
