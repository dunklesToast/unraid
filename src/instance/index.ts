export * as executor from './executor';
export * as unraid from './unraid';
