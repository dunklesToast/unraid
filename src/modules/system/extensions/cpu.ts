import type { RecursivePartial } from '../../../util/type';
import { SystemModuleExtensionBase } from '../system-module-extension-base';

export interface CoreUsage {
    /**
     * time running un-niced user processes
     */
    usr: number;
    /**
     * time running niced user processes
     */
    nice: number;
    /**
     * time running kernel processes
     */
    sys: number;
    /**
     * time waiting for I/O completion
     */
    ioWait: number;
    /**
     * time spent in the kernel idle handler
     */
    idle: number;
    /**
     * time spent servicing hardware interrupts
     */
    hardwareInterrupts: number;
    /**
     * time spent servicing software interrupts
     */
    softwareInterrupts: number;
    /**
     * time stolen from this vm by the hypervisor
     */
    steal: number;
}

export interface CPUUsage {
    coreCount: number;
    all: CoreUsage;
    cores: (CoreUsage & { core: number })[];
    raw: string[];
}

export type ICpuUsageStreamOption = {
    /**
     * Refresh time in seconds
     */
    refresh?: number;
};

export class SystemModuleCpuExtension extends SystemModuleExtensionBase {
    private static cpuUsageCommand = `TERM=xterm top -1 -n 1 -b | grep '^%Cpu[[:digit:]+]' | tr '\n' '|'`;

    private static cpuLineRegex =
        /%Cpu(?<cpu_core>\d+)\s+:\s+(?<user>\d+.\d+(?=\s+us)).*,\s*(?<system>\d+.\d+(?=\s+sy)).*,\s*(?<nice>\d+.\d+(?=\s+ni)).*,\s*(?<idle>\d+.\d+(?=\s+id)).*,\s*(?<io_wait>\d+.\d+(?=\s+wa)).*,\s*(?<hardware_interrupts>\d+.\d+(?=\s+hi)).*,\s*(?<software_interrupts>\d+.\d+(?=\s+si)).*,\s*(?<steal>\d+.\d+(?=\s+st))/gm;

    async usage(): Promise<CPUUsage> {
        const { code, stdout } = await this.instance.execute(SystemModuleCpuExtension.cpuUsageCommand);
        if (code !== 0) throw new Error('Got non-zero exit code while getting cpu usage');

        return SystemModuleCpuExtension.parseCpuUsageOutput(stdout);
    }

    private static parseCpuUsageOutput(input: string[]): CPUUsage {
        const tempUsage: RecursivePartial<CPUUsage> = {
            all: {
                usr: 0,
                sys: 0,
                nice: 0,
                hardwareInterrupts: 0,
                idle: 0,
                ioWait: 0,
                softwareInterrupts: 0,
                steal: 0,
            } as CoreUsage,
            cores: [],
            raw: [...input],
        };

        // %Cpu0  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
        // %Cpu7  :  0.0 us,  5.0 sy,  0.0 ni, 95.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
        const coreLines = input
            .join('')
            .split('|')
            .filter((v) => v);

        for (const line of coreLines) {
            const match = SystemModuleCpuExtension.cpuLineRegex.exec(line);
            const core = parseInt(match.groups.cpu_core, 10);
            const user = parseFloat(match.groups.user);
            const system = parseFloat(match.groups.system);
            const nice = parseFloat(match.groups.nice);
            const idle = parseFloat(match.groups.idle);
            const ioWait = parseFloat(match.groups.io_wait);
            const hardwareInterrupts = parseFloat(match.groups.hardware_interrupts);
            const softwareInterrupts = parseFloat(match.groups.software_interrupts);
            const steal = parseFloat(match.groups.steal);
            tempUsage.all.usr += user;
            tempUsage.all.sys += system;
            tempUsage.all.nice += nice;
            tempUsage.all.idle += idle;
            tempUsage.all.ioWait += ioWait;
            tempUsage.all.hardwareInterrupts += hardwareInterrupts;
            tempUsage.all.softwareInterrupts += softwareInterrupts;
            tempUsage.all.steal += steal;
            tempUsage.cores.push({
                core,
                usr: user,
                sys: system,
                nice,
                idle,
                ioWait,
                hardwareInterrupts,
                softwareInterrupts,
                steal,
            });
            SystemModuleCpuExtension.cpuLineRegex.lastIndex = 0;
        }

        tempUsage.all.usr /= tempUsage.cores.length;
        tempUsage.all.sys /= tempUsage.cores.length;
        tempUsage.all.nice /= tempUsage.cores.length;
        tempUsage.all.idle /= tempUsage.cores.length;
        tempUsage.all.ioWait /= tempUsage.cores.length;
        tempUsage.all.hardwareInterrupts /= tempUsage.cores.length;
        tempUsage.all.softwareInterrupts /= tempUsage.cores.length;
        tempUsage.all.steal /= tempUsage.cores.length;

        return tempUsage as CPUUsage;
    }

    on_cpuUsage(listener: (newLoad: CPUUsage) => void, options?: ICpuUsageStreamOption): [() => void] {
        let cancelFunction;

        if (!this.instance.executor.executeStream) {
            throw new Error('Streaming is not supported by this executor');
        }

        const refresh = options?.refresh ?? 1;

        const promise = this.instance.executor
            .executeStream(
                `while [ 1 ]; do ${SystemModuleCpuExtension.cpuUsageCommand} | xargs -0 echo; sleep ${refresh}; done`
            )
            .then(([eventEmitter, cancel, exitPromise]) => {
                cancelFunction = () => {
                    cancel().finally(() => {
                        eventEmitter.removeAllListeners('onNewStdoutLine');
                    });
                };
                eventEmitter.addListener('onNewStdoutLine', (line) => {
                    listener(SystemModuleCpuExtension.parseCpuUsageOutput([line]));
                });
                exitPromise.finally(() => {
                    eventEmitter.removeAllListeners('onNewStdoutLine');
                });
            });

        return [
            () => {
                promise.finally(() => {
                    if (cancelFunction) {
                        cancelFunction();
                    }
                });
            },
        ];
    }
}
