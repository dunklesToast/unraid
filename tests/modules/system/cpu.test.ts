import '@system/cpu';
import { Unraid } from '../../../src';
import { SSHConfig, TestExecutor } from '../../executor/TestExecutor';
import { getInstance } from '../../helpers/getInstance';

describe('system - cpu', () => {
  let instance: Unraid<SSHConfig, TestExecutor>;

  beforeAll(() => {
    instance = getInstance();
    global.mock = 'cpu';
  });

  it('usage', async () => {
    const usage = await instance.system.usage();
    expect(usage).toMatchSnapshot();
  });
});
