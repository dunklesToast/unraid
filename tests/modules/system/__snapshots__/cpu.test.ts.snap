// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`system - cpu usage 1`] = `
Object {
  "all": Object {
    "hardwareInterrupts": 0,
    "idle": 56.31666666666667,
    "ioWait": 30.954166666666662,
    "nice": 4,
    "softwareInterrupts": 0,
    "steal": 0,
    "sys": 5.875,
    "usr": 2.8791666666666664,
  },
  "cores": Array [
    Object {
      "core": 0,
      "hardwareInterrupts": 0,
      "idle": 94.4,
      "ioWait": 0,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 5.6,
      "usr": 0,
    },
    Object {
      "core": 1,
      "hardwareInterrupts": 0,
      "idle": 85,
      "ioWait": 10,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 5,
      "usr": 0,
    },
    Object {
      "core": 2,
      "hardwareInterrupts": 0,
      "idle": 84.2,
      "ioWait": 5.3,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 10.5,
      "usr": 0,
    },
    Object {
      "core": 3,
      "hardwareInterrupts": 0,
      "idle": 60,
      "ioWait": 30,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 10,
      "usr": 0,
    },
    Object {
      "core": 4,
      "hardwareInterrupts": 0,
      "idle": 70,
      "ioWait": 15,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 10,
      "usr": 5,
    },
    Object {
      "core": 5,
      "hardwareInterrupts": 0,
      "idle": 23.8,
      "ioWait": 52.4,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 4.8,
      "usr": 19,
    },
    Object {
      "core": 6,
      "hardwareInterrupts": 0,
      "idle": 0,
      "ioWait": 78.9,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 15.8,
      "usr": 5.3,
    },
    Object {
      "core": 7,
      "hardwareInterrupts": 0,
      "idle": 81,
      "ioWait": 14.3,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 0,
      "usr": 4.8,
    },
    Object {
      "core": 8,
      "hardwareInterrupts": 0,
      "idle": 55,
      "ioWait": 35,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 0,
      "usr": 10,
    },
    Object {
      "core": 9,
      "hardwareInterrupts": 0,
      "idle": 68.4,
      "ioWait": 26.3,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 0,
      "usr": 5.3,
    },
    Object {
      "core": 10,
      "hardwareInterrupts": 0,
      "idle": 0,
      "ioWait": 94.7,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 5.3,
      "usr": 0,
    },
    Object {
      "core": 11,
      "hardwareInterrupts": 0,
      "idle": 81,
      "ioWait": 9.5,
      "nice": 4.8,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 4.8,
      "usr": 0,
    },
    Object {
      "core": 12,
      "hardwareInterrupts": 0,
      "idle": 65,
      "ioWait": 25,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 10,
      "usr": 0,
    },
    Object {
      "core": 13,
      "hardwareInterrupts": 0,
      "idle": 77.8,
      "ioWait": 22.2,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 0,
      "usr": 0,
    },
    Object {
      "core": 14,
      "hardwareInterrupts": 0,
      "idle": 89.5,
      "ioWait": 0,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 5.3,
      "usr": 5.3,
    },
    Object {
      "core": 15,
      "hardwareInterrupts": 0,
      "idle": 0,
      "ioWait": 85.7,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 9.5,
      "usr": 4.8,
    },
    Object {
      "core": 16,
      "hardwareInterrupts": 0,
      "idle": 76.2,
      "ioWait": 4.8,
      "nice": 9.5,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 4.8,
      "usr": 4.8,
    },
    Object {
      "core": 17,
      "hardwareInterrupts": 0,
      "idle": 80,
      "ioWait": 5,
      "nice": 15,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 0,
      "usr": 0,
    },
    Object {
      "core": 18,
      "hardwareInterrupts": 0,
      "idle": 0,
      "ioWait": 85,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 15,
      "usr": 0,
    },
    Object {
      "core": 19,
      "hardwareInterrupts": 0,
      "idle": 66.7,
      "ioWait": 0,
      "nice": 28.6,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 4.8,
      "usr": 0,
    },
    Object {
      "core": 20,
      "hardwareInterrupts": 0,
      "idle": 65,
      "ioWait": 25,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 10,
      "usr": 0,
    },
    Object {
      "core": 21,
      "hardwareInterrupts": 0,
      "idle": 0,
      "ioWait": 95,
      "nice": 0,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 5,
      "usr": 0,
    },
    Object {
      "core": 22,
      "hardwareInterrupts": 0,
      "idle": 38.1,
      "ioWait": 23.8,
      "nice": 28.6,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 4.8,
      "usr": 4.8,
    },
    Object {
      "core": 23,
      "hardwareInterrupts": 0,
      "idle": 90.5,
      "ioWait": 0,
      "nice": 9.5,
      "softwareInterrupts": 0,
      "steal": 0,
      "sys": 0,
      "usr": 0,
    },
  ],
  "raw": Array [
    "%Cpu0  :  0.0 us,  5.6 sy,  0.0 ni, 94.4 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu1  :  0.0 us,  5.0 sy,  0.0 ni, 85.0 id, 10.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu2  :  0.0 us, 10.5 sy,  0.0 ni, 84.2 id,  5.3 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu3  :  0.0 us, 10.0 sy,  0.0 ni, 60.0 id, 30.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu4  :  5.0 us, 10.0 sy,  0.0 ni, 70.0 id, 15.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu5  : 19.0 us,  4.8 sy,  0.0 ni, 23.8 id, 52.4 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu6  :  5.3 us, 15.8 sy,  0.0 ni,  0.0 id, 78.9 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu7  :  4.8 us,  0.0 sy,  0.0 ni, 81.0 id, 14.3 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu8  : 10.0 us,  0.0 sy,  0.0 ni, 55.0 id, 35.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu9  :  5.3 us,  0.0 sy,  0.0 ni, 68.4 id, 26.3 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu10 :  0.0 us,  5.3 sy,  0.0 ni,  0.0 id, 94.7 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu11 :  0.0 us,  4.8 sy,  4.8 ni, 81.0 id,  9.5 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu12 :  0.0 us, 10.0 sy,  0.0 ni, 65.0 id, 25.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu13 :  0.0 us,  0.0 sy,  0.0 ni, 77.8 id, 22.2 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu14 :  5.3 us,  5.3 sy,  0.0 ni, 89.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu15 :  4.8 us,  9.5 sy,  0.0 ni,  0.0 id, 85.7 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu16 :  4.8 us,  4.8 sy,  9.5 ni, 76.2 id,  4.8 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu17 :  0.0 us,  0.0 sy, 15.0 ni, 80.0 id,  5.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu18 :  0.0 us, 15.0 sy,  0.0 ni,  0.0 id, 85.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu19 :  0.0 us,  4.8 sy, 28.6 ni, 66.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu20 :  0.0 us, 10.0 sy,  0.0 ni, 65.0 id, 25.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu21 :  0.0 us,  5.0 sy,  0.0 ni,  0.0 id, 95.0 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu22 :  4.8 us,  4.8 sy, 28.6 ni, 38.1 id, 23.8 wa,  0.0 hi,  0.0 si,  0.0 st|%Cpu23 :  0.0 us,  0.0 sy,  9.5 ni, 90.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st|",
  ],
}
`;
